### ISSUE TITLE
Feature request: Large unicode range on windows platform, devel build.

### ISSUE THREAD

**Issue by jburse** [SWI-Prolog/swipl-devel#354]
Would it be possible to have a SWI-Prolog .exe on the Windows platform
with a larger character range. On the Mac platform the range is larger,
but not on the Windows platform. Or do I miss something, is there a 
start-up option or a special .exe to enable a larger range on Windows platform?

![grafik](https://user-images.githubusercontent.com/5411487/47618032-8ab47300-dace-11e8-8922-6399be0e2666.png)

**Comment by jburse** [comment]
With the limitations it is difficult to replicate SWI-Prolog Mac stuff on 
SWI-Prolog windows. But such a larger range seems to be more and 
more common. Like Tau Prolog on the browser platform can already
do from a few days ago. See for example:  

https://github.com/jariazavalverde/tau-prolog/issues/56

**Comment by JanWielemaker** [comment]
Should probably be on the roadmap repo.  Not very likely to happen before there is someone with interest in this and either willing to implement it or willing to pay someone to do so.  Roughly there are two options:

  - Ditch using `wchar_t` and use UTF-8 for everything internally.
  - Use a new type throughout for wide characters internally that is uint32_t.  This maps to `wchar_t`
    on non-Windows.  On Windows we must replace the C99 wide character functions with functions
    that work on this new type.   There might be some external library doing that.

Both require significant work.

**Comment by jburse** [comment]
Currently thinking about something related to this feature request. Like a ascii(true) option for write_term/2. So that although if Unicode is present, like on the Mac platform, that things could be written out in ascii common denominator format.

As a next step, then maybe make write_term/2, write/1, writeq/1, etc.. more intelligent, by querying the underlying stream for a default value of the new ascii option. Since it happens on the Windows platform, that not all consoles can also accept unicode.

**Comment by JanWielemaker** [comment]
That is taken care of.  You can set the encoding of a stream and what to do if a character doesn't fit in the encoding.

**Comment by jburse** [comment]
You refer to this comment https://github.com/SWI-Prolog/swipl-devel/issues/354#issuecomment-434008359, where I suggested a related feature, not the original feature request. And you closed the original feature request claiming the related feature works.

But the related feature works neither. Was testing on the Mac Platform just now:

![foo_failed](https://user-images.githubusercontent.com/5411487/47722885-a85f1500-dc53-11e8-8b72-a9844b9b6072.png)

But the related feature, how I specified it, adapted to your encoding/1 option, would automatically switch to a lower representation, and write simply '\x1F600\' to the output stream. I guess SWI-Prolog does not yet implement such a feature.
